src/swarm/neo/client/Connection.d

```diff
@@ -331,6 +331,19 @@ public final class Connection: ConnectionBase
 
     /***************************************************************************
 
+        Causes the connection to be dropped. It will be re-established as
+        normal. This method is only intended for use in tests.
+
+    ***************************************************************************/
+
+    public void reconnect ( )
+    {
+        this.socket.shutdown();
+        this.status_ = this.status_.Disconnected;
```

Comment:
@david-eckardt-sociomantic would mind confirming that this is enough for the reconnect process to start? I think it is, but I would like to double check if we're not missing something.

Reply:
Looks good to me but I think `socket.close` should be used instead. `close` implies `shutdown`, and `shutdown` is useful only in special cases such as shutting the connection down in only one direction. Here you really want to `close` the socket, you cannot reuse it after `shutdown` anyway but create a new socket.

Reply:
This code is the result of experiment without really looking deeply into what's going on.

I tried using `close`, but was surprised to find that it didn't work. I can try it again and report what the problem is.

Reply:
Interesting... it seems that `close` makes the send/receive loops exit, which essentially makes the connection unusable. `shutdown` somehow does not have this effect.

Reply:
I believe the difference lies somewhere here (behavior of the `recv` on `shutdown` vs `close`)

```
       These calls return the number of bytes received, or -1 if an error
       occurred.  In the event of an error, errno is set to indicate the
       error.

       When a stream socket peer has performed an orderly shutdown, the
       return value will be 0 (the traditional "end-of-file" return).
```

http://man7.org/linux/man-pages/man2/recv.2.html

Reply:
(iow. `socket.close` causes reads to report error, while `socket.shutdown` will report feof (0 bytes read)).

Reply:
Fair enough, so there is a good reason to use `shutdown`.

Reply:
Maybe I'll document this.

Reply:
Debug test output:
```
ConnectionSet: reconnect [127, 0, 0, 1]:10000
Connection fd=17 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: Connection fd=17 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: Connection fd=18 events=ReRh
Connection fd=17 events=ReRh :: Epoll firing with events ReHuRh
SendLoop.fiberMethod() caught "Expected non-zero errno after failure (connection hung up on read)" @./src/swarm/neo/protocol/socket/MessageReceiver.d:558
Connection fd=17 events=ReRh :: Unregistered
[127, 0, 0, 1]:10000: Connection.connect()
[127, 0, 0, 1]:10000: Connection.tryConnect()
[127, 0, 0, 1]:10000: ClientConnect.connect()
ConnectProtocol fd=5 events=Wr :: Registered
ReceiveLoop.fiberMethod() exit
Connection fd=5 events= :: Handled, leaving fd registered
Connection fd=18 events=ReRh :: Epoll firing with events ReRh
SendLoop.fiberMethod() caught "Expected non-zero errno after failure (connection hung up on read)" @./src/swarm/neo/protocol/socket/MessageReceiver.d:558
node connection shutdown "Expected non-zero errno after failure (connection hung up on read)" @./src/swarm/neo/protocol/socket/MessageReceiver.d:558
Connection fd=18 events=ReRh :: Unregistered
node connection shutdown success
SendLoop.fiberMethod() exit
ReceiveLoop.fiberMethod() exit
Connection fd=-1 events= :: Handled, leaving fd registered
EpollSelectDispatcher.select (5 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=Wr
     4: YieldedRequestOnConns fd=15 events=Re
SelectListener fd=13 events=Re :: Epoll firing with events Re
SendLoop.fiberMethod() 54996
ConnectProtocol fd=6 events=ReRh :: Registered
SelectListener fd=13 events=Re :: Handled, leaving fd registered
ConnectProtocol fd=5 events=Wr :: Epoll firing with events Wr
[127, 0, 0, 1]:10000: ClientConnect.connect() succeeded
[127, 0, 0, 1]:10000: ClientConnect.autenticate()
[127, 0, 0, 1]:10000: Auth: Sent client protocol version = 1
ConnectProtocol fd=5 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: ConnectProtocol fd=6 events=ReRh
ConnectProtocol fd=6 events=ReRh :: Epoll firing with events Re
Read client protocol version = 1
Sent node protocol version = 1
Protocol versions match
ConnectProtocol fd=6 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: ConnectProtocol fd=6 events=ReRh
ConnectProtocol fd=5 events=ReRh :: Epoll firing with events Re
[127, 0, 0, 1]:10000: Auth: Read node protocol version = 1
[127, 0, 0, 1]:10000: Auth: Protocol versions match
[127, 0, 0, 1]:10000: Auth: Timestamp=1508242264
ConnectProtocol fd=5 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: ConnectProtocol fd=6 events=ReRh
ConnectProtocol fd=6 events=ReRh :: Epoll firing with events Re
ConnectProtocol fd=6 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: ConnectProtocol fd=6 events=ReRh
ConnectProtocol fd=5 events=ReRh :: Epoll firing with events Re
[127, 0, 0, 1]:10000: Auth: Nonce=[168, 64, 111, 61], Name=dummy, Encoded=[89, EB, C4, 58, ED, 71, 7D, 30, 2A, 68, 76, EC, 7F, 7A, 38, DC, 2A, 7B, C6, 55, 90, 98, E8, C8, 8B, DD, 27, 52, C7, 45, 39, 87, 6C, 96, 10, 61, DD, E0, 83, 56, BC, 81, 5B, B6, 68, 55, C0, 42, DD, 80, DB, 6D, D4, 4E, 1A, B, 97, D6, 1A, 8C, 19, C7, 1D, 97]
ConnectProtocol fd=5 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: ConnectProtocol fd=6 events=ReRh
ConnectProtocol fd=6 events=ReRh :: Epoll firing with events Re
SwarmConn: Client timestamp=1508242264, Node timestamp=1508242264
SwarmAuth: Confirm encoded=[89, EB, C4, 58, ED, 71, 7D, 30, 2A, 68, 76, EC, 7F, 7A, 38, DC, 2A, 7B, C6, 55, 90, 98, E8, C8, 8B, DD, 27, 52, C7, 45, 39, 87, 6C, 96, 10, 61, DD, E0, 83, 56, BC, 81, 5B, B6, 68, 55, C0, 42, DD, 80, DB, 6D, D4, 4E, 1A, B, 97, D6, 1A, 8C, 19, C7, 1D, 97]
ConnectProtocol fd=6 events=ReRh :: Unregistered
SendLoop.fiberMethod() connected
ReceiveLoop.fiberMethod()
Connection fd=6 events=ReRh :: Registered
ConnectProtocol fd=6 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: ConnectProtocol fd=5 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: Connection fd=6 events=ReRh
ConnectProtocol fd=5 events=ReRh :: Epoll firing with events Re
[127, 0, 0, 1]:10000: ClientConnect.autenticate() succeeded
ConnectProtocol fd=5 events=ReRh :: Unregistered
[127, 0, 0, 1]:10000: ClientConnect.connect() exit
[127, 0, 0, 1]:10000: Connection.tryConnect() succeeded
[127, 0, 0, 1]:10000: Connection.tryConnect() exit
```

Debug test output with `close` instead:
```
ConnectionSet: reconnect [127, 0, 0, 1]:10000
Connection fd=-1 events=ReRh :: Handled, leaving fd registered
EpollSelectDispatcher.select (6 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: Connection fd=-1 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
     5: Connection fd=18 events=ReRh
Connection fd=18 events=ReRh :: Epoll firing with events ReRh
SendLoop.fiberMethod() caught "Expected non-zero errno after failure (connection hung up on read)" @./src/swarm/neo/protocol/socket/MessageReceiver.d:558
node connection shutdown "Expected non-zero errno after failure (connection hung up on read)" @./src/swarm/neo/protocol/socket/MessageReceiver.d:558
Connection fd=18 events=ReRh :: Unregistered
node connection shutdown success
SendLoop.fiberMethod() exit
ReceiveLoop.fiberMethod() exit
Connection fd=-1 events= :: Handled, leaving fd registered
EpollSelectDispatcher.select (5 clients registered):
     0: YieldedRequestOnConns fd=12 events=Re
     1: SelectListener fd=14 events=Re
     2: SelectListener fd=13 events=Re
     3: Connection fd=-1 events=ReRh
     4: YieldedRequestOnConns fd=15 events=Re
<HANGS>
```

Reply:
Updated with an explanation of why `shutdown` is what we want here.